### ISSUE TITLE
RPC Client isn't compatible with app.route()

### ISSUE THREAD

**Issue by Soulusions** [honojs/hono#2120]
### What version of Hono are you using?

3.12.8

### What runtime/platform is your app running on?

Bun

### What steps can reproduce the bug?

```ts
// routes/routerA.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { z } from "zod";

const app = new Hono();

app.get("/", (c) => {
  return c.json({ message: "Hello" }, 200);
});

app.get(
  "/:id",
  zValidator(
    "param",
    z.object(
      {
        id: z.string()
      })
  ),
  async (c) => {
    return c.json({ message: `Hello ${c.req.param("id")}` }, 200);
  });

export default app;

// index.ts
import { Hono } from "hono";
import { default as routerA } from "./routes/routerA";
import { hc } from "hono/client";

const app = new Hono();

const routes = app.route("/a", routerA);

export default app;
export type RpcType = typeof routes;

const rpc = hc<RpcType>("http://localhost:3000/");
            ^ unknown
```

### What is the expected behavior?

The `rpc` object should have proper typings.

### What do you see instead?

The `rpc` object doesn't have typings.

**Comment by Soulusions** [comment]
I've found a very janky, incomplete workaround, by doing
```ts
// routes/routerA.ts
import { Hono } from "hono";
import { zValidator } from "@hono/zod-validator";
import { z } from "zod";

const app = new Hono();

const route1 = app.get("/", (c) => {
  return c.json({ message: "Hello" }, 200);
});

const route2 = app.get(
  "/:id",
  zValidator(
    "param",
    z.object(
      {
        id: z.string()
      })
  ),
  async (c) => {
    return c.json({ message: `Hello ${c.req.param("id")}` }, 200);
  });

export default app;
export type RouteARpc = typeof route1 & typeof route2;
```
This "works", but trying to & types from various routers, they are obviously unaware of the fact their base path is changed in the index.ts file, so this method isn't really worthwhile...

**Comment by Soulusions** [comment]
Ohhh, I see, how would that work if I added a B router, so something the [building a larger application](https://hono.dev/guides/best-practices#building-a-larger-application) example, since I think that's what a lot of people might end up doing and it doesn't seem really documented how that'd work with the RPC client.

**Comment by yusukebe** [comment]
Or, and this may be a matter of taste, it will be correctly inferred with chaining.

```ts
// routes/routerA.ts
import { zValidator } from '@hono/zod-validator'
import { Hono } from 'hono'
import { z } from 'zod'

const app = new Hono()
  .get('/', (c) => {
    return c.json({ message: 'Hello' }, 200)
  })
  .get(
    '/:id',
    zValidator(
      'param',
      z.object({
        id: z.string(),
      })
    ),
    async (c) => {
      return c.json({ message: `Hello ${c.req.param('id')}` }, 200)
    }
  )

export default app
```

```ts
// index.ts
import { Hono } from 'hono'
import { hc } from 'hono/client'
import { default as routerA } from './routerA'

const app = new Hono()

const routes = app.route('/a', routerA)

export type RpcType = typeof routes
const rpc = hc<RpcType>('http://localhost:3000/')
```

**Comment by yusukebe** [comment]
I would really like to achieve this using the method you described at the beginning, but the limitations of TypeScript prevent us from doing so. So we only have the two methods I described above.

> Ohhh, I see, how would that work if I added a B router, so something the [building a larger application](https://hono.dev/guides/best-practices#building-a-larger-application) example, since I think that's what a lot of people might end up doing and it doesn't seem really documented how that'd work with the RPC client.

Ah, you're right, we should write it down.

**Comment by Soulusions** [comment]
Thank you very much for the help, I managed to get things working by simply using chaining, on both the "controllers" and the `.route()` calls in the `index.ts` file, I think it's not a bad experience, simply that since the RPC client only ever really explains how to handle 1 route, it's not very clear chaining is required!
Though I also understand the limitations of typescript and the fact it would be hard to achieve what I first did without very complex types that would slow down the TS LSP down.
Thank you very much for your help!
